securedrop/alembic/versions/17c559a7a685_pgp_public_keys.py

```diff
@@ -0,0 +1,92 @@
+"""PGP public keys
+
+Revision ID: 17c559a7a685
+Revises: 811334d7105f
+Create Date: 2023-09-21 12:33:56.550634
+
+"""
+
+import traceback
+
+import pretty_bad_protocol as gnupg
+import sqlalchemy as sa
+from alembic import op
+from encryption import EncryptionManager
+from sdconfig import SecureDropConfig
+
+# revision identifiers, used by Alembic.
+revision = "17c559a7a685"
+down_revision = "811334d7105f"
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    """
+    Migrate public keys from the GPG keyring into the SQLite database
+
+    We iterate over all the secret keys in the keyring, identify the
+    corresponding source, export the key and save it in the database.
+    """
+    config = SecureDropConfig.get_current()
+    gpg = gnupg.GPG(
+        binary="gpg2",
+        homedir=str(config.GPG_KEY_DIR),
+        options=["--pinentry-mode loopback", "--trust-model direct"],
+    )
+    # Source keys all have a secret key, so we can filter on that
+    for keyinfo in gpg.list_keys(secret=True):
+        if len(keyinfo["uids"]) > 1:
+            # Source keys should only have one UID
+            continue
+        uid = keyinfo["uids"][0]
+        search = EncryptionManager.SOURCE_KEY_UID_RE.search(uid)
+        if not search:
+            # Didn't match at all
+            continue
+        filesystem_id = search.group(2)
+        # Check that it's a valid ID
+        conn = op.get_bind()
+        result = conn.execute(
+            sa.text(
+                """
+                SELECT pgp_public_key, pgp_fingerprint
+                FROM sources
+                WHERE filesystem_id=:filesystem_id
+                """
+            ).bindparams(filesystem_id=filesystem_id)
+        ).first()
+        if result != (None, None):
+            # Either not in the database or there's already some data in the DB.
+            # In any case, skip.
+            continue
+        fingerprint = keyinfo["fingerprint"]
+        try:
+            public_key = gpg.export_keys(fingerprint)
+        except:  # noqa: E722
+            traceback.print_exc()
+            continue
```

Comment:
If we think this actually might fail on a production instance, then we should consider:
1. What state do we want the instance to be in?
2. What action, if any, do we want the administrator to (have to) take?

Concretely, even if we want the rest of the migration to continue, do we want the upgrade to exit nonzero and require some kind of recovery intervention before it can ultimately pass?

I suspect the answer is probably no, and I'm fine with that.  But I'm asking now rather than waiting for v2.7 QA.  :-)

Reply:
I changed my mind on this twice in the middle of this PR, so here's where I ended up, and what the code implements.

Our primary goal with Sequoiafication is to improve stability and reliability, especially as we look to primarily focus on SDW. This requires that the migration process also be robust, especially in the face of potential GPG errors that we're trying to eliminate going forwards.

So what happens if the migration fails with a non-zero exit code? dpkg will error out, meaning that some updates may be interrupted. It's not clear to me if it blocks further updates of other packages, but having an incompletely upgraded package just seems like a bad spot to be in because we can't assert or reason about what state the instance is in.

What if we had the migration complete with a zero exit code, but set some flag as failed for admin intervention? We would then have to ask the maintainer to run some `gpg` commands manually, inspecting the keyring for that specific source (or sources). I think this the opposite of what we want, because it means we have to spend time debugging gpg instead of forgetting about it!

So then we're left with our current implementation of doing the migration, but quietly failing on errors. This means that the migration will optimistically happen for each source, but if there are issues it'll be skipped on a case-by-case basis. The GPG fallback still exists in `Source.public_key`, so if it was a spurious failure of some kind, the source will still be operable and can receive replies, etc.

I took special care to not increase our footprint of pretty_bad_protocol (see also my attempt at tree shaking it: https://github.com/freedomofpress/securedrop/pull/6960), all the functions used are also used elsewhere:
* list_keys() - EncryptionManager._get_source_key_details()
* export_keys() - EncryptionManager._get_public_key()

This pre-existing usage informed how I chose to guard them - list_keys() is called for every source and also in startup in prime_keycache(), so if it was broken, the whole instance would be busted, and therefore I think it's safe to assume it's successful. But if export_key() was broken, you'd have to be looking at that specific source to know (prime_keycache would theoretically hit it, but it silently suppresses errors), so it's possible some very old source is in this case, but no one ever triggered it, and the migration would uncover it.

And because the usage of pbp is the same as the migration itself, there's very little to be gained (IMO) by removing the GPG fallbacks, versus the upside of the migration being buggy somehow but everything still working.

Hopefully that was not too rambley and makes sense. Also maybe there's somewhere better to document this design decision besides a CR comment?

Reply:
Thanks for laying out your thinking here, @legoktm, and for implementing this so carefully.  I'm on board.  For me this is the fundamental decision:

> So then we're left with our current implementation of doing the migration, but quietly failing on errors. This means that the migration will optimistically happen for each source, but if there are issues it'll be skipped on a case-by-case basis. The GPG fallback still exists in `Source.public_key`, so if it was a spurious failure of some kind, the source will still be operable and can receive replies, etc.

I agree that this is worth highlighting and recording, so I've summarized it in <https://github.com/freedomofpress/securedrop/wiki/2.7.0-development-notes>.  Feel free to amend (or move elsewhere) as you see fit.